python/GafferUI/SpreadsheetUI/_ProxySelectionModel.py

```diff
@@ -0,0 +1,175 @@
+		self.__settingSelection = False
+
+		QtCore.QItemSelectionModel.__init__( self, viewModel, parent )
+
+		self.setTargetSelectionModel( targetSelectionModel )
+		self.__initMappings()
+
+		self.currentChanged.connect( self.__currentChanged )
+		self.modelChanged.connect( self.__modelChanged )
+
+	def setTargetSelectionModel( self, selectionModel ) :
+
+		if selectionModel == self.__targetSelectionModel :
+			return
+
+		if self.__targetSelectionModel :
+			self.disconnect( self.__targetSelectionModel )
+
+		self.__targetSelectionModel = selectionModel
+
+		self.__initMappings()
+
+		self.__targetSelectionModel.selectionChanged.connect( self.__sourceSelectionChanged )
+		self.__targetSelectionModel.currentChanged.connect( self.__sourceCurrentChanged )
+		self.__targetSelectionModel.modelChanged.connect( self.__modelChanged )
+
+	def getTargetSelectionModel( self ) :
+
+		return self.__targetSelectionModel
+
+	def select( self, selection, flags ) :
+
+		assert( isinstance( selection, ( QtCore.QItemSelection, QtCore.QModelIndex ) ) )
+
+		if isinstance( selection, QtCore.QModelIndex ) :
+			# The overload of this just wraps the selection and calls back here
+			# which results in double evaluation
+			selection = QtCore.QItemSelection( selection, selection )
+
+		QtCore.QItemSelectionModel.select( self, selection, flags )
+
+		# Forward to source model
+		self.__settingSelection = True
+		sourceSelection = self.mapSelectionToSource( selection )
+		self.__targetSelectionModel.select( sourceSelection, flags )
+		self.__settingSelection = False
+
+	def mapToSource( self, index ) :
+
+		assert( self.__modelChain )
+
+		for model in self.__modelChain :
+			index = model.mapToSource( index )
+
+		if index.isValid() :
+			assert( index.model() is self.__targetSelectionModel.model() )
+
+		return index
+
+	def mapFromSource( self, index ) :
+
+		assert( self.__modelChain )
+
+		for model in reversed( self.__modelChain ) :
+			index = model.mapFromSource( index )
+
+		if index.isValid() :
+			assert( index.model() is self.model() )
+
+		return index
+
+	def mapSelectionToSource( self, itemSelection ) :
+
+		assert( self.__modelChain )
+
+		for model in self.__modelChain :
+			itemSelection = model.mapSelectionToSource( itemSelection )
+
+		return itemSelection
+
+	def mapSelectionFromSource( self, itemSelection ) :
+
+		assert( self.__modelChain )
+
+		for model in reversed( self.__modelChain ) :
+			itemSelection = model.mapSelectionFromSource( itemSelection )
+
+		return itemSelection
+
+	def __modelChanged( self ) :
+
+		self.__initMappings()
+
+	def __initMappings( self ) :
+
+		# Build a chain of models from this views model to the target
+
+		assert( self.model() )
+		assert( self.__targetSelectionModel )
+
+		model = self.model()
+		targetModel = self.__targetSelectionModel.model()
+
+		models = []
+		while model != targetModel :
+			assert( hasattr( model, 'sourceModel' ) )
+			models.append( model )
+			model = model.sourceModel()
```

Comment:
Do we expect to find more than one here? It looks like we do, but I haven't understood why.

Reply:
Potentially. Just thought I'd writing it properly to start with, so later on when we add filters, etc.. we don't spend ages wondering why selection isn't working. This class isn't specific to the spreadsheet UI so could easily be used in other places later if we need.